v1.0: Primary key inference (#2104)
meilisearch/documentation: learn/core_concepts/primary_key.md

### Meilisearch guesses your primary key

Suppose you add documents to an index without previously setting its primary key. In this case, Meilisearch will automatically look for an attribute ending with the string `id` in a case-insensitive manner (for example, `uid`, `BookId`, `ID`) in your first document and set it as the index's primary key.

> If Meilisearch finds multiple attributes ending with `id`, it will throw the [`index_primary_key_multiple_candidates_found`](#index-primary-key-multiple-candidates-found) error and ask you to manually set the primary key. If it cannot find a suitable attribute, it will throw the [`index_primary_key_no_candidate_found`](#index-primary-key-no-candidate-found) error. In both cases, the document addition process will be interrupted and no documents will be added to your index.

Reviewer: I felt that the error code names are unnecessary here, since they make the text more congested and are also mentioned further down on the page. Please push back if you disagree.

```suggestion
If Meilisearch finds [multiple attributes ending with `id`](#index-primary-key-multiple-candidates-found) or [cannot find a suitable attribute](#index-primary-key-no-candidate-found), it will throw an error. In both cases, the document addition process will be interrupted and no documents will be added to your index.
```

## Primary key errors

This section covers some primary key errors and how to resolve them.

### `index_primary_key_no_candidate_found`

> This error occurs when you add documents for the first time and none of them have an attribute ending with `id`. It can be resolved by [manually setting the index's primary key](#setting-the-primary-key-on-document-addition), or ensuring that all documents you add possess an `id` attribute.

Reviewer: ```suggestion
This error occurs when you add documents to an index for the first time and none of them have an attribute ending with `id`. It can be resolved by [manually setting the index's primary key](#setting-the-primary-key-on-document-addition), or ensuring that all documents you add possess an `id` attribute.
```

```json
{
  "uid": 1,
  "indexUid": "books",
  "status": "failed",
  "type": "documentAdditionOrUpdate",
  "canceledBy": null,
  "details":{

>     "receivedDocuments":5,
>     "indexedDocuments":5

Author: ```suggestion
    "receivedDocuments": 5,
    "indexedDocuments": null
```

  },
  "error":{

>     "message": "The primary key inference process failed because the engine did not find any field ending with `id` in its name. Please specify the primary key manually using the `primaryKey` query parameter.",

Reviewer 2: Ah, sorry for not noticing this earlier, but it looks like the error message is not the latest version we decided upon together (the [specification PR](https://github.com/meilisearch/specifications/pull/208) is [up to date](https://github.com/meilisearch/specifications/pull/208#issuecomment-1372005125))

```suggestion
    "message": "The primary key inference failed as the engine did not find any field ending with `id` in its name. Please specify the primary key manually using the `primaryKey` query parameter.",
```

### Meilisearch guesses your primary key

> Suppose you add documents to an index without previously setting its primary key. In this case, Meilisearch will automatically look for an attribute ending with the string `id` in a case-insensitive manner (for example, `uid`, `BookId`, `ID`) in your first document and set it as the index's primary key. If it finds multiple attributes ending with `id`, it will throw an error and ask you to manually set the primary key.

Reviewer 2: I wonder if we could describe the error case with multiple attributes maybe in the next paragraph as that would emphasize that the behavior of Meilisearch is similar when there is no candidate and when there are multiple candidates found.

> If Meilisearch cannot find a suitable attribute, or if it finds multiple attributes, ...

    "finishedAt": "2021-12-30T11:28:59.084803Z"
}
```

> ### `index_primary_key_multiple_candidates_found`

Reviewer: Should this follow directly the other new error, since they both have to do with primary key inference?

Author: It would be better to list these errors in alphabetical order as we do in `error_codes.md`

Reviewer: I disagree.

Alphabetical order makes sense for a reference file like `error_codes.md`, where people will be searching for a specific error. Being alphabetical helps them find that error more quickly.

Unlike `error_codes.md`, this page is intended to be read sequentially. Because of that, these errors should be grouped by topic/conceptually.

Reviewer: In this case, it seems like alphabetical order puts the errors in an intuitive order anyway, so further changes may not be required. Just something to think about in the future.

> This error occurs when Meilisearch fails to guess your primary key as the index has multiple attributes containing `id`. It can be resolved by [manually setting the index's primary key](#setting-the-primary-key-on-document-addition).

Reviewer: Do you see why I am suggesting these changes?

```suggestion
This error occurs when you add documents to an index for the first time and Meilisearch finds multiple attributes ending with `id`. It can be resolved by [manually setting the index's primary key](#setting-the-primary-key-on-document-addition).
```

Author: The sentence paraphrases the error and isn't useful

Reviewer: The main thing is that the difference between "containing id" and "ending with id" is potentially huge; it's the difference between having an accurate understanding and expectation of the inference, and not.

Reviewer: The other thing is that saying "the index has multiple attributes containing `id`" is incorrect and potentially confusing. When the error occurs, the document addition fails, so the index is empty.

```json
{
  "uid": 4,
  "indexUid": "books",
  "status": "failed",
  "type": "documentAdditionOrUpdate",
  "canceledBy":null,
  "details":{
    "receivedDocuments":5,
    "indexedDocuments":5
  },
  "error":{

>     "message": "The primary key inference process failed because the engine found 2 fields ending with `id` in their name, such as 'id' and 'author_id'. Please specify the primary key manually using the `primaryKey` query parameter.",

Reviewer 2: Same here:

```suggestion
    "message": "The primary key inference failed as the engine found 2 fields ending with `id` in their names: 'id' and 'author_id'. Please specify the primary key manually using the `primaryKey` query parameter.",
```
